---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    highlight: zenburn
    theme: flatly
---

```{R, message=F, warning=F}
library(dplyr)
library(biomaRt)
library(tximport)
library(rhdf5)
library(gplots)
library(DESeq2)
library(DT)
library(apeglm)
library(RColorBrewer)
library(IHW)
library(PCAtools)
library(pheatmap)
library(EnhancedVolcano)
library(ComplexHeatmap)
library(circlize)
library(fgsea)
library(tidyverse)
library(ggpubr)
```

# Introduction
This R markdown document walks through an RNA-Seq differential expression analysis using `DESeq2` for MA5112. Quality control plots, differential expression results, plots and pathway analysis are covered in the document. 

The tutorial assumes you have 9 Kallisto quantification directories under `~/RNA-Seq/quant/` on your laptop. 

# Reading Input Data

### Metadata
Read in the file `samples.csv` which contains the experimental metadata. Please make sure that the rownames of this dataframe match the kallisto quantification directory names at `/home/RNA-Seq/quant/`. 
```{R}
samples <- read.csv("~/RNA-Seq/samples.csv", header=T, row.names = "sample", stringsAsFactors = T)
samples
```


### Convert Numerics to Factors
When performing differential expression analysis using `DESeq2`, if a column in `samples.csv` is numeric and provided to the design formula, `DESeq2` will produce the following message:

> the design formula contains a numeric variable with integer values, specifying a model with increasing fold change for higher values. did you mean for this to be a factor? if so, first convert this variable to a factor using the factor() function

What does this mean? From Mike Love (`DESeq2` author): 

> There is a constant fold change for every unit of change in replicates. So if the estimated fold change is 2, this implies that replicates 2 = 2x replicates 1, replicates 3 = 2x replicates 2, etc. Or in other words, the relationship is linear on the log counts scale. 

This is not what we want. We want each replicate to be modelled as a seperate term in the model. To overcome this, we could convert 1 -> A, 2 -> B, 3 -> C or we can use `factor()` to specify that replicates should be treated as a factor, not a numeric.

The code block below checks the input metadata file, tells the user if any columns are numeric and converts them to a factor.
```{R, message=F, warning=F}
## check if all columns are factors.  
factor_cols <- sapply(samples, is.factor)

if(all(factor_cols) == TRUE){
  print("All columns in metadata are factors and suitable for analysis.")
}else{
  numeric_cols <- sapply(samples, is.numeric)
  names <- colnames(samples)[numeric_cols]
  print(paste0("Column(s) ", names, " is numeric. Converting to factor."))
  samples[numeric_cols] <- as.data.frame(lapply(samples[numeric_cols], factor))
  final_check <- sapply(samples, is.factor)
  if(all(final_check) == TRUE){
    print("All columns in metadata are factors and suitable for analysis.")
  }else{
    print("Error in converting to factors. Check input metadata file.")
  }
}
```

### Stage Kallisto files
We need to create a file handle object (a named character list) containing the sample IDs and the paths to the kallisto quantification `.h5` files. If the rownames of your metadata object (`samples`) do not match the quantification directory names, you will get an error during the `TXI object` step.
```{R, message=F, warning=F}
dir <- ("~/RNA-Seq/quant")
files <- file.path(dir, rownames(samples), "abundance.h5")
names(files) <- paste0(rownames(samples))
files
```

# Tximport
`tximport` imports transcript-level abundances from quantification tools (`kallisto` in our case) and converts them to gene counts for downstream analyses using a differential expression analysis package. 

### BiomaRt
We will use `biomaRt` to connect to the `ENSEMBL` databases and map transcript IDs to gene IDs. Firstly, we will create the `mart` object specifying which database to use:
```{R, warning=F, message=F}
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
```

### Transcript to gene (tx2gene)
Recall we used the `ENSEMBL` reference cDNA file `Homo_sapiens.GRCh38.cdna.all.fa.gz` for `Kallisto quant` on LUGH. This means that our quantification files have `ENSEMBL` transcript ID's. We can map them to gene symbols running the code below: 
```{R, message=F, warning=F}
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version", "hgnc_symbol"), mart = mart, useCache = FALSE)
head(tx2gene)
```

### TXI object
Create a `txi` object summarising `kallisto` transcipt quantification to the gene-level. 
```{R, message=F, warning=F}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene)
head(txi$abundance)
```

# DESeq2
Now we are ready to create the `DDS` object for `DESeq2` analysis that contains metadata (colData), counts further information that can be obtained via `dds@`.

### Design 
The model design is one of the most important steps for an RNA-Seq analysis. Here, we are going to specify the design `~ replicate + condition` (columns of the metadata file). The factor of interest goes last in the model terms, we want to compare control vs. lung vs. melanoma in the experiment. Replicate has been included as we want to control for the effect of sample replicates.

### DDS
We are going to create the `DDS` object using `DESeqDataSetFromTximport()` as we used `tximport` to convert transcript abundances to gene-level counts.
```{R, message=F, warning=F}
dds <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ replicate + condition )
```

### Factor levels
By default, R will choose a reference level for factors based on alphabetical order. If you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. Set the reference level using `relevel()`, and create the `(DESeq) DDS` object using `DESeq()`.

Check that `relevel()` worked correctly by running `resultsNames(dds)` - we want control to be the reference level so we can compare lung vs control and melanoma vs. control. 
```{R, message=F, warning=F}
dds$condition <- relevel(dds$condition, ref = "control") 
dds <- DESeq(dds) 
resultsNames(dds)
```

# Quality Control 
We will use the gene-level counts to perform some quality control checks on the samples in the experiment. 

### Extract Counts
Extract gene-level counts from the `DDS` object.
```{R, message=F, warning=F}
counts <- counts(dds, normalized=TRUE)
log2 <- log2(counts + 1)
```

### Save Counts
Write the counts files to text files in `~/RNA-Seq`
```{R, message=F, warning=F}
dir.create("~/RNA-Seq/counts")
write.table(counts, "~/RNA-Seq/counts/normalised_counts.txt", sep="\t", quote = F)
write.table(log2, "~/RNA-Seq/counts/log2_counts.txt", sep="\t", quote = F)
```

### Sample Heatmap
A heatmap of sample distances matrix (`dist(t(log2))`) gives us an overview over similarities and dissimilarities between samples. 
```{R, message=F, warning=F}
sampleDists <- dist(t(log2))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- NULL
pheatmap(mat=sampleDistMatrix,
            clustering_distance_rows=sampleDists,
            clustering_distance_cols=sampleDists,
            col=colorRampPalette( rev(brewer.pal(9, "Blues")) )(255))

```

### PCA
PCA plots the variance explained by samples in each principal component. Typically PC1 & PC2 explain the most variation in the dataset, you should look at these first and foremost. 

In the plot below we can see that the lung and control samples cluster tightly together, indicating the replicates were performed rigorously in the lab. 

The melanoma samples overlap with the control samples in PC1/PC2. Make the effort to explore other principal components - we can see that in PC2 vs PC4 there is excellent separation between the melanoma and control samples. 

PCA can give you a good idea of how succesful the Differential Expression analysis will be - samples that are very close together in PC feature space will not produce as many DE genes as those that are separated by a large distance.
```{R, message=F,warning=F}
p <- pca(log2, metadata = samples)

biplot(p,
       colby = 'condition', 
       colkey = c('melanoma'='royalblue', 'control'='red1', 'lung'='forestgreen'),
       hline = 0, 
       vline = 0,
       legendPosition = 'right', 
       legendLabSize = 12, 
       legendIconSize = 8.0,
       lab = TRUE,
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC2')

biplot(p,
       x = "PC2",
       y = "PC4",
       colby = 'condition', 
       colkey = c('melanoma'='royalblue', 'control'='red1', 'lung'='forestgreen'),
       hline = 0, 
       vline = 0,
       legendPosition = 'right', 
       legendLabSize = 12, 
       legendIconSize = 8.0,
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC2 versus PC4')
```

# Differential Expression Analysis

### DESeq results()
We can extract differentially expressed genes between phenotypes of interest by using `results()` on the `DDS` object. Furtehrmore, we will apply the `apeglm` shrinkage estimator on our results. I will demonstrate the effect of `apeglm` with volcano plots. In summary `apeglm` shrinks low confidence (high inter sample variation) differentially expressed genes towards 0, producing a robust set of differentially expressed genes.

Set up lung vs control, melanoma vs control:
```{R, message=F, warning=F}
# make lung vs control object
lung_v_ctrl <- results(dds, filterFun=ihw, alpha=0.05, c("condition", "lung", "control"))
res1 <- lfcShrink(dds=dds, res=lung_v_ctrl, coef=4, type="apeglm")
summary(res1)

# make melanoma vs control object 
melanoma_v_ctrl <- results(dds, filterFun=ihw, alpha=0.05, c("condition", "melanoma", "control"))
res2 <- lfcShrink(dds=dds, res=melanoma_v_ctrl, coef=5, type="apeglm")
summary(res2)
```

Set up lung vs melanoma:
```{R, message=F, warning=F}
# to make lung vs melanoma, relevel the dds object reference level and redo the DESeq call
dds$condition <- relevel(dds$condition, ref = "melanoma")
dds <- DESeq(dds)

# double chekck it worked
resultsNames(dds)

# make lung vs melanoma 
lung_v_melanoma <- results(dds, filterFun=ihw, alpha=0.05, c("condition", "lung", "melanoma"))
res3 <- lfcShrink(dds=dds, res=lung_v_melanoma, coef=5, type="apeglm")
summary(res3)
```

> Take a look at the summary statistics of the number of DE genes. Do they conform to comments made about the PCA plot? 

### Resusable Functions
Below are some handy functions I re-use all the time for extracting differentially expressed genes and performing annotation of DE genes. 

##### Extract Up Regulated Genes
```{R}
get_upregulated <- function(df){

	key <- intersect(rownames(df)[which(df$log2FoldChange>=1)], rownames(df)[which(df$pvalue<=0.05)])
  
  results <- as.data.frame((df)[which(rownames(df) %in% key),])
	return(results)
}
```

##### Extract Down Regulated genes
```{R}
get_downregulated <- function(df){

  	key <- intersect(rownames(df)[which(df$log2FoldChange<=-1)], rownames(df)[which(df$pvalue<=0.05)])
  
  	results <- as.data.frame((df)[which(rownames(df) %in% key),])
  	return(results)
}
```

##### Annotate DE genes
```{R}
annotate_de_genes <- function(df){

    df$hgnc_symbol <- rownames(df)
    mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
    info <- getBM(attributes=c("hgnc_symbol",
                               "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand",
                               "entrezgene_description"),
                  filters = c("hgnc_symbol"),
                  values = df$hgnc_symbol,
                  mart = mart,
                  useCache=FALSE)

    tmp <- merge(df, info, by="hgnc_symbol")
    tmp$strand <- gsub("-1", "-", tmp$strand)
    tmp$strand <- gsub("1", "+", tmp$strand)
    tmp$hgnc_symbol <- make.names(tmp$hgnc_symbol, unique = T)
    tmp <- tmp[!grepl("CHR", tmp$chromosome_name),]

    output_col <- c("Gene", "Ensembl ID", "Chromosome", "Start", "Stop", "Strand", "Description", "Log2FC", "P-value", "Adj P-value")
    tmp <- subset(tmp, select=c(hgnc_symbol, ensembl_gene_id_version, chromosome_name, start_position, end_position, strand, entrezgene_description, log2FoldChange, pvalue, padj))
    colnames(tmp) <- output_col

    if(min(tmp$Log2FC) > 0){
        tmp <- tmp[order(-tmp$Log2FC),]
    }else{
        tmp <- tmp[order(tmp$Log2FC),]
    }

    return(tmp)

}
```


### Write DE results
Use the above functions to extract up/down regulated genes and annotate them using biomaRt, and write to a file. Please note that the get_(up/down)regulated function requires a dataframe as input.
```{R, message=F, warning=F}
de_up <- get_upregulated(as.data.frame(res1))
de_down <- get_downregulated(as.data.frame(res1))

upregulated_genes <- annotate_de_genes(de_up)
downregulated_genes <- annotate_de_genes(de_down)

dir.create("~/RNA-Seq/DESeq_results/")
write.table(upregulated_genes, "~/RNA-Seq/DESeq_results/lung_vs_control_upregulated.txt", sep="\t", row.names=F, quote=F)	
write.table(downregulated_genes, "~/RNA-Seq/DESeq_results/lung_vs_control_downregulated.txt", sep="\t", row.names=F, quote=F)	
```

# Results Plots

### Volcano Plot
Volcano plots are useful to show how many genes are differentially expressed in the experimental contrast of interest. Labels are optional, I have included them so you know how to use them.

N.B Volcano plots use -log10 on the Y-axis.

```{R, message=F, warning=F}
res1 <- na.omit(res1)
	
min_width <- min(res1$log2FoldChange)
max_width <- max(res1$log2FoldChange)
max_height <- -log10(min(res1[res1$pvalue>0, 5]))

up <- subset(res1, res1$log2FoldChange > 1 & res1$pvalue <= 0.05)
up <- up[order(-up$log2FoldChange),]
up_list <- head(rownames(up), n=10L)

down <- subset(res1, res1$log2FoldChange < -1 & res1$pvalue <= 0.05)
down <- down[order(down$log2FoldChange),]
down_list <- head(rownames(down), n=10L)

plot_top_20 <- c(up_list, down_list)
EnhancedVolcano(res1,
			          lab=rownames(res1),
          			x="log2FoldChange",
          			y="pvalue",
          			selectLab=plot_top_20,
          			drawConnectors=TRUE,
          			FCcutoff=1.0,
          			pCutoff=0.05,
          			title="Volcano Plot",
          			subtitle="Lung vs. Control",
          			legendVisible=F,
          			caption = paste0('Total Genes = ', nrow(res1)),
          			xlim=c(min_width, max_width),
          			ylim=c(0, max_height))
```

### Bonus: apeglm visualised
Make a volcano plot of `lung_v_ctrl`, the `DESeq2` results object that has not been filtered by `apeglm`. 

You can see that there are genes with very large log2 FC values. As a rule of thumb, be skeptical of genes that have LFC values over +/- 10. 
```{R, message=F, warning=F}
res1 <- na.omit(lung_v_ctrl)
	
min_width <- min(res1$log2FoldChange)
max_width <- max(res1$log2FoldChange)
max_height <- -log10(min(res1[res1$pvalue>0, 5]))

up <- subset(res1, res1$log2FoldChange > 1 & res1$pvalue <= 0.05)
up <- up[order(-up$log2FoldChange),]
up_list <- head(rownames(up), n=10L)

down <- subset(res1, res1$log2FoldChange < -1 & res1$pvalue <= 0.05)
down <- down[order(down$log2FoldChange),]
down_list <- head(rownames(down), n=10L)

plot_top_20 <- c(up_list, down_list)
EnhancedVolcano(res1,
			          lab=rownames(res1),
          			x="log2FoldChange",
          			y="pvalue",
          			selectLab=plot_top_20,
          			drawConnectors=TRUE,
          			FCcutoff=1.0,
          			pCutoff=0.05,
          			title="Volcano Plot",
          			subtitle="Lung vs. Control",
          			legendVisible=F,
          			caption = paste0('Total Genes = ', nrow(res1)))

##reset res1 for heatmaps
res1 <- na.omit(res1)
```

### Complex Heatmap
Heatmaps are another way to show the differentially expressed genes in the experimental contrast of interest. Complexheatmap is as the name suggest, quite complex compared to other heatmap packages but it is undoubtedly the best in terms of flexibility.
```{R, message=F, warning=F}
# subset the counts matrix to get the lung and control samples
subset <- counts[, 1:6]

# now select de_up, de_down, i.e DE genes that passed the filtering
up <- rownames(de_up)
down <- rownames(de_down)

# subset them
key <- c(up, down)
subset <- subset[which(rownames(subset) %in% key),]

# Scale and center the counts matrix
# Scale works on columns, transpose the matrix to scale and center genes, transpose back.
mat <- t(subset)
mat <- scale(mat, center=T, scale=T)
mat <- t(mat)
mat <- na.omit(mat)

# set up annotation dataframe
ann <- data.frame(Cell_Type = c(rep("control", 3), c(rep("lung", 3))))

# set up heatmap column annotation
ha_col = HeatmapAnnotation(df = ann, 
                           col = list(Cell_Type = c("control" =  "gold", 
                                                    "lung" = "forestgreen")), 
                           annotation_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 12)), 
                          annotation_name_side = "left")


# check the min and max values for our color gradient
max(mat)
min(mat)


# set up heatmap object
hm1 <- Heatmap(mat, 
              col= colorRamp2(c(-2.6,-1,0,1,2.6),c("blue","skyblue","white","lightcoral","red")),
              heatmap_legend_param=list(at=c(-2.6,-1,0,1,2.6),color_bar="continuous", 
                                        legend_direction="vertical", legend_width=unit(5,"cm"),
                                        title_position="topcenter", title_gp=gpar(fontsize=10, fontface="bold")), 
              name = "Z-score",
              
              #Row annotation configurations
              cluster_rows=T,
              show_row_dend=T,
              row_title_side="right",
              row_title_gp=gpar(fontsize=8),
              show_row_names=FALSE,
              row_names_side="left",
              
              #Column annotation configuratiions
              cluster_columns=T,
              show_column_dend=T,
              column_title="Lung vs. Control DE Genes",
              column_title_side="top",
              column_title_gp=gpar(fontsize=15, fontface="bold"),
              show_column_names = T,
              column_names_gp = gpar(fontsize = 12, fontface="bold"),
              
              #Dendrogram configurations: columns
              clustering_distance_columns="euclidean",
              clustering_method_columns="complete",
              column_dend_height=unit(10,"mm"),

              #Dendrogram configurations: rows
              clustering_distance_rows="euclidean",
              clustering_method_rows="complete",
              row_dend_width=unit(4,"cm"),
              row_dend_side = "left",
              row_dend_reorder = TRUE,

              #Splits
              border=T,
              row_km = 1,
              column_km = 1,
              
              #plot params
              width = unit(5, "inch"), 
              height = unit(4, "inch"),
              #height = unit(0.4, "cm")*nrow(mat),
              
              #Annotations
              top_annotation = ha_col)

# plot heatmap
draw(hm1, annotation_legend_side = "right", heatmap_legend_side="right")
```

### Heatmap (top 20 genes)
Instead of plotting all DE genes in a heatmap, we can select the top n differentially expressed genes to plot. I have decided to use the largest LFC values to plot the top 20 genes, however you might see people plotting the top n variable genes in this situation. 
```{R}
# same logic as volcano plot
up <- subset(de_up, de_up$log2FoldChange > 1 & de_up$pvalue <= 0.05)
up <- up[order(-up$log2FoldChange),]
up_list <- head(rownames(up), n=10L)

down <- subset(de_down, de_down$log2FoldChange < 1 & de_down$pvalue <= 0.05)
down <- down[order(down$log2FoldChange),]
down_list <- head(rownames(down), n=10L)

plot_top_20 <- c(up_list, down_list)

# subset the counts matrix from last step
top_genes <- mat[which(rownames(mat) %in% plot_top_20),]

hm1 <- Heatmap(top_genes, 
              col= colorRamp2(c(-2.6,-1,0,1,2.6),c("blue","skyblue","white","lightcoral","red")),
              heatmap_legend_param=list(at=c(-2.6,-1,0,1,2.6),color_bar="continuous", 
                                        legend_direction="vertical", legend_width=unit(5,"cm"),
                                        title_position="topcenter", title_gp=gpar(fontsize=10, fontface="bold")), 
              name = "Z-score",
              
              #Row annotation configurations
              cluster_rows=T,
              show_row_dend=T,
              row_title_side="right",
              row_title_gp=gpar(fontsize=8),
              show_row_names=TRUE,
              row_names_side="right",
              
              #Column annotation configuratiions
              cluster_columns=T,
              show_column_dend=T,
              column_title="Lung vs. Control top 20 DE genes",
              column_title_side="top",
              column_title_gp=gpar(fontsize=15, fontface="bold"),
              show_column_names = T,
              column_names_gp = gpar(fontsize = 12, fontface="bold"),
              
              #Dendrogram configurations: columns
              clustering_distance_columns="euclidean",
              clustering_method_columns="complete",
              column_dend_height=unit(10,"mm"),

              #Dendrogram configurations: rows
              clustering_distance_rows="euclidean",
              clustering_method_rows="complete",
              row_dend_width=unit(4,"cm"),
              row_dend_side = "left",
              row_dend_reorder = TRUE,

              #Splits
              border=T,
              row_km = 1,
              column_km = 1,
              
              #plot params
              width = unit(5, "inch"), 
              height = unit(4, "inch"),
              #height = unit(0.4, "cm")*nrow(mat),
              
              #Annotations
              top_annotation = ha_col)

# plot heatmap
draw(hm1, annotation_legend_side = "right", heatmap_legend_side="right")
```

# Pathway Analysis
Please download the GO Biological processes GMT file from the following link: https://raw.githubusercontent.com/BarryDigby/barrydigby.github.io/master/RNA-Seq/c5.bp.v7.0.symbols.gmt

Run fGSEA:
```{r, message=F, warning=F}
## convert result object to dataframe
res <- as.data.frame(res1)
res$hgnc_symbol <- rownames(res)

## use absolute value for LFC 
res$log2FoldChange <- abs(res$log2FoldChange)

# compute summary stat
fgsea_rank <- res %>%
              dplyr::select(hgnc_symbol, log2FoldChange) %>%
              na.omit() %>%
              distinct() %>%
              group_by(hgnc_symbol) %>%
              summarize(stat=mean(log2FoldChange))

# create named list
rank <- deframe(fgsea_rank)

# read in gmt file
pathway <- gmtPathways("~/RNA-Seq/c5.bp.v7.0.symbols.gmt")

# run fgsea
fgsea <- fgseaMultilevel(pathways=pathway, stats=rank, eps = 0, scoreType = "pos") %>%
                         as_tibble() %>%
                         arrange(padj)

head(fgsea)
```

### Enrichment plots
Subset the `fgsea` results object to include results whose adjusted pvalue is < 0.0001. You can use any preferred filtering method here, this is just to return a handful of pathways for the enrichment plot for loop. 
```{R, message=F, warning=F}
filtered_pathway <- subset(fgsea, padj < 0.0001)

filt_p <- as.vector(filtered_pathway$pathway)

for (i in filt_p){
    plt <- plotEnrichment(pathway = pathway[[i]], 
    gseaParam = 1, ticksSize = 0.5, stats= rank) + 
    labs(title=i) + theme(plot.title = element_text(hjust = 0.5, face="bold"))
    print(plt)
}
```

