---
title: Kallisto bustools Pipeline
layout: page
permalink: /scRNA-Seq/pipeline
---

A description of the `kallisto bustools` pipeline has been provided for you below.

There is no need to run this for this weeks tutorial, it has been posted as a learning resource.

1. [Indexing](#index)
2. [kallisto bus](#bus)
3. [bustools correct](#correct)
4. [bustools sort](#sort)
5. [bustools text](#text)

***

# 1. Indexing {#index}
Kallisto requires an indexed genome file for downstream quantification.

##### *Inputs*
- Reference Genome

```bash
kallisto index -i GRCh37.idx GRCh37.fa
```

##### *Flags*
* `-i`: Filename for the kallisto index to be constructed

##### *Outputs*
Kallisto index produces an indexed genome file with the filename specified by `-i`

***

# 2. Kallisto bus {#bus}
Generate BUS files for single-cell sequencing

##### *Inputs*
- Genome Index file
- FASTQ reads
- 10x Genomics Chemistry version (string)

```bash
kallisto bus \
         -i Genome Index file \
         -o outDir/ \
         -x 10xv3 \
         -t 2 \
         FASTQ_1, FASTQ_2
```

##### *Flags*
* `-i` Genome Index file generated by kallisto `index`
* `-o` Directory to write output to
* `-x` Single-cell technology used
* `-t` *n* threads to use

##### *Outputs*
Kallisto `bus` generates an output directory with the following files:

- `matrix.ec`
- `output.bus`
- `run_info.json`
- `transcripts.txt`

***

# 3. Bustools correct {#correct}
BUS files can be barcode error corrected w.r.t. a technology specific whitelist of barcodes. The correct command will correct all barcodes that are at Hamming distance 1 (i.e. one substitution) away from a single barcode in the whitelist.

##### *Inputs*
- Whitelisted barcodes file
- BUS file generated by kallisto `bus`

```bash
bustools correct \
         -w whitelist.file \
         output.bus \
         -o output.corrected.bus
```

##### *Flags*
- `-o` File for corrected bus output
- `-w` File of whitelisted barcodes to correct to

##### *Outputs*
Bustools `correct` outputs a corrected BUS file.

***

# 4. Bustools sort {#sort}
Raw BUS output from pseudoalignment programs may be unsorted. To simply and accelerate downstream processing BUS files can be sorted using bustools sort.

##### *Inputs*
- corrected BUS file generated by `bustools correct`

```bash
bustools sort \
         -T tmp/ \
         -t 2 \
         -o output.sort.bus \
         output.corrected.bus
```

##### *Flags*
- `-t` Number of threads to use
- `-m` Maximum memory used
- `-T` Location and prefix for temporary files
- `-o` File for sorted output

##### *Outputs*
Bustools `sort` outputs a sorted BUS file.

***

# 5. Bustools text {#text}
BUS files can be converted to a tab-separated format for easy inspection and processing using shell scripts or high level languages before importing to R/Python for downstream analysis.

##### *Inputs*
- Corrected + Sorted BUS file
- Transcripts to Gene mapping file

```bash
bustools text \
         -o output.sort.txt \
         output.sort.bus
```

##### *Flags*
- `-o` File for text output

##### *Outputs*
Bustools `text` produces a BUS file in text format.
