---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    highlight: zenburn
    theme: flatly
---

# MA5112 scRNA-Seq

```{R, warning=F, message=F}
library(Seurat)
library(patchwork)
library(dplyr)
```

## Load Data

Establish path to directory containing outputs from `Kallisto Bustools`:
```{R, message=F, warning=F}
pbmc.data <- Read10X(data.dir = "/data/github/barrydigby.github.io/scRNA-Seq/seurat/")
```

**Create Seurat Object:**
Filtering can be applied in this step, where we can select:

* `min.features`: Include cells where at least this many features are detected. 
* `min.cells`: Include features detected in at least this many cells. 

For the tutorial we will select conservative filtering values:
```{R, message=F, warning=F}
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc1k", min.cells = 2, min.features = 100)
```

### Seurat Object 

Inspect the `pbmc` object by using a `$` or `@` symbol to access RNA counts/features and metadata, respectively. 

Check the number of genes and samples (cells) in the experiment: 
```{R}
pbmc
```
One can check the `nCount_RNA` and `nFeature_RNA` values per cell by using `pbmc$nFeature_RNA` and `pbmc$nFeature_RNA`. Alternatively, use `pbmc@meta.data` to obtain an overview of both: 
```{R, message=F, warning=F}
head(pbmc@meta.data)
```

### Count Matrix

In contrast to RNA-Seq where empty cells are represented with a `0`, scRNA-seq use `.` in place of `0` (sparse matrix representation) to significantly reduce memory costs. The code below shows the count matrix for three genes over the first 10 cells: 
```{R, message=F, warning=F}
pbmc.data[c("TRDD2", "DR1", "TARBP1"), 1:10]
```

## Pre-processing

We have already performed a pre-processing step using `CreateSeruatObject()`, whereby cells with low numbers of genes expressed and lowly expressed genes were filtered using `min.features` and `min.cells`, respectively. 

Further common filtering steps include: 

* Removing cells with aberrantly high gene counts
* Removing low quality/dying cells (such cells exhibit mitochondrial contamination)

### Mitochondrial Genes

Calculate the fraction of mitochondrial genes expressed in each cell using `PercentageFeatureSet()` and REGEX, appending the results to the `meta.data` of the seurat object:
```{R, message=F, warning=F}
pbmc[["percent_mt"]] = PercentageFeatureSet(pbmc, pattern="^MT-")
```

Inspect the results:
```{R, message=F, warning=F}
head(pbmc@meta.data)
```

Let's make a plot of these statistics to inform cut-off values:
```{R, message=F, warning=F}
VlnPlot(pbmc, features = c("nCount_RNA", "nFeature_RNA", "percent_mt"), ncol = 3)
```

Based on the plots, perform filtering to remove:

* Cells with mitochondrial gene `%` above 10% 
* Cells with gene counts above 20,000
* Cells with unique feature counts above 6500 and below 300

```{R, message=F, warning=F}
pbmc <- subset(pbmc, subset = nFeature_RNA > 300 & nFeature_RNA < 6500 & nCount_RNA < 20000 & percent_mt < 10)
```

Plot the filtered object: 
```{R, message=F, warning=F}
VlnPlot(pbmc, features = c("nCount_RNA", "nFeature_RNA", "percent_mt"), ncol = 3)
```

## Normalization 

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in `pbmc[["RNA"]]@data`:
```{R, message=F, warning=F}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```

## Feature Selection

In order to extract meaningful biological signals from the dataset, Seurat aims to identify a subset of features (e.g., genes) exhibiting high variability across cells, and therefore represent heterogeneous features to prioritize for downstream analysis. Choosing genes solely based on their log-normalized single-cell variance fails to account for the mean-variance relationship that is inherent to single-cell RNA-seq. Therefore, a variance-stabilizing transformation is applied to correct for this before calculating the mean-variance relationship, implemented in the `FindVariableFeatures()` function. 
```{R, message=F, warning=F}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)
```
### Plot variable genes

Plot the top 20 variable genes: 
```{R, message=F, warning=F}
top20 <- head(VariableFeatures(pbmc), 20)
init_plot <- VariableFeaturePlot(pbmc)
plot_var <- LabelPoints(plot = init_plot, points = top20, repel = TRUE)
plot_var
```

## Data Scaling

The dataset now needs to be centered and scaed prior to performing dimensionality reduction techniques such as `PCA`. To reiterate what we mean by scaling: 

* Shift the expression of each gene, so that the mean expression across cells is 0
* Scale the expression of each gene, so that the variance across cells is 1
   * This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
   
Perform scaling, the results are stored in `pbmc[["RNA"]]@scale.data`:
```{R, message=F, warning=F}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

## Dimension Reduction

### PCA

Perform `PCA` on the selected features in the `pbmc` object. If you wish to provide a different assay to `PCA`, define it using `features`:
```{R, message=F, warning=F}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

### Inspect Loadings
Inspect the loadings on each PC:
```{R, message=F, warning=F}
print(pbmc[["pca"]], dims = 1:2, nfeatures = 5)
```
### Plot Loadings

Plot the loadings on each PC (note the plots seem to rank features according to absolute values, hence why negative genes are dominating the plots):
```{R, message=F, warning=F, fig.width=8}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca", nfeatures = 5)
```

### Plot 2D Scatter

Plot a 2D scatter plot of each cell whereby the cells position is determined by the embeddings determined by the reduction technique:
```{R, message=F, warning=F}
DimPlot(pbmc, reduction = "pca")
```

### Heatmap of loadings
 
`DimHeatmap()` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores.
```{R, message=F, warning=F}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

## Choose PCs

The reason we performed PCA is to compress the dataset into a robust representation of the heterogeneity present in the dataset for downstream clustering, however now we are faced with an issue: how many PCs to include for downstream analyses? 

### JackStraw

Seurat implements a resampling test inspired by the `JackStraw` procedure. Seurat randomly permutes a subset of the data (1% by default) and reruns PCA, constructing a ‘null distribution’ of feature scores, and repeats the procedure. Seurat then identifies ‘significant’ PCs as those who have a strong enrichment of low p-value features:
```{R, message=F, warning=F}
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:15)
```

### Plot JackStraw

The `JackStrawPlot()` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 7 PCs. 

```{R, message=F, warning=F}
JackStrawPlot(pbmc, dims = 1:15)
```

### Elbow Plot

An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each PC. In the below plot we might conclude that taking the the top 10 PCs makes the most sense. 

> Ask Sarah which method is most appropriate, should significant PCs in JackStraw be included?

```{R, message=F, warning=F}
ElbowPlot(pbmc)
```

## Cluster Cells

Seurat applies a graph-based clustering approach by first constructing a KNN graph based on the euclidean distance in PCA space, and refining the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the `FindNeighbors()` function, and takes as input the previously defined dimensionality of the dataset (we will include the top 10 PCs here).

To cluster the cells, Seurat next applies modularity optimization techniques such as the Louvain algorithm (default) or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. The `FindClusters()` function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets:
```{R, message=FALSE, warning=FALSE}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

The clusters can be found using the `Idents()` function:
```{R, message=F, warning=F}
head(Idents(pbmc), 5)
```

## UMAP/tSNE

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{R, message=F, warning=F}
pbmc <- RunUMAP(pbmc, dims = 1:10)
DimPlot(pbmc, reduction = "umap")
```

```{R, message=F, warning=F}
pbmc <- RunTSNE(pbmc, dims = 1:10)

DimPlot(pbmc, reduction = "tsne")
```

## Differentially Expressed features

Seurat can help you find markers that define clusters via differential expression. `FindAllMarkers()` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

### FindMarkers
In the example below we will test the differentially expressed features of cluster 0, specifying that the genes must be expressed in 25% of both cell groups:
```{R, message=F, warning=F}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, min.pct = 0.25)
head(cluster0.markers, n = 5)
```

In the example below we will test differentially expressed features of cluster 1 vs cluster 0 & 4, with the same percentage cut-off:
```{R, message=F, warning=F}
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, ident.2 = c(0, 4), min.pct = 0.25)
head(cluster1.markers, n = 5)
```

### FindAllMarkers

In the example below we test differentially expressed markers in each cluster vs all other cells iteratively:
```{R, message=F, warning=F}
pbmc.markers <- FindAllMarkers(pbmc, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```

### Violin Plot

Plot individual gene expression counts across clusters:
```{R, message=F, warning=F}
VlnPlot(pbmc, features = c("IGKC", "GNLY"), slot = "counts")
```

### Features Plot

Plot differentially expressed features in the reduced dimensional embedding plot (PCA, UMAP, tSNE), set `interactive = TRUE` for tutorial. 
```{R, message=F, warning=F}
FeaturePlot(pbmc, features = c( "GNLY", "IGKC"))#, interactive = TRUE)
```

### Heatmap

```{R, message=F, warning=F, fig.width=8, fig.height=10}
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```